Added plugin Plabric

plabric opened the pull request.

foosel commented:
Hi there! I just took a look at your plugin and there are a couple of critical points to address with it:

  * Your plugin extracts the global API key with admin rights & apparently sends it somewhere (your docker container?). I'd like to know where, and why. There are probably better ways to go about this, it feels VERY wrong and potentially puts the user into harm's way.
  * I see references to Docker. Please explain in what way this plugin depends on Docker and why it apparently can't run without. It feels like WAY too much overhead just for what claims to be a VPN connection, especially if full access to the OctoPrint instance is shared with the container. Additionally I cannot find the image in question anywhere.
  * Installing Docker for the user *even with permission* is not something I want to see a registered plugin doing, especially not if it then installs arbitrary unknown containers into said Docker instance. Hands off of system packages, there's SSH for this kind of stuff!
  * Same goes for creating folders somewhere in the user's file system outside of the plugin folder ("docker_img").
  * Do NOT modify bundled KO bindings like `qrcode`, that will cause core OctoPrint and/or third party plugins to malfunction in really hard to debug ways!
  * Your plugin is not Python 3 compatible. Python 2 is EOL. OctoPrint 1.4.0 does support Python 2 and Python 3, plugins will be expected to do so as well (or at least support Python 3) during a transition period of roughly a year after the release of OctoPrint 1.4.0 to allow a migration of the whole ecosystem. Newly registered plugins should be Python 3 compatible out of the box. [See this forum thread for more information](https://community.octoprint.org/t/towards-python-3-and-octoprint-1-4-0/12382?u=foosel).

Can you please look into fixing these and then ping me for another review here once it's done? Thank you!

plabric commented:
Hi @foosel,

Thank you for your time.

> * Your plugin extracts the global API key with admin rights & apparently sends it somewhere (your docker container?). I'd like to know where, and why. There are probably better ways to go about this, it feels VERY wrong and potentially puts the user into harm's way.

The plugin sends the API key to the Docker container to allow access to the Octoprint API. But for what your comment indicates, it's not the right way to do it. Please, could you tell me whatwould be the correct way to do it? Thanks in advance.

> * I see references to Docker. Please explain in what way this plugin depends on Docker and why it apparently can't run without. It feels like WAY too much overhead just for what claims to be a VPN connection, especially if full access to the OctoPrint instance is shared with the container. Additionally I cannot find the image in question anywhere.

The docker image opens a WebRTC connection to Plabric App using [node-webrtc](https://github.com/node-webrtc/node-webrtc) and makes signaling with Plabric Server using [sockjs-client](https://github.com/sockjs/sockjs-client).
The docker image is downloaded after plugin installation when user access to plugin settings and click the button to install the plugin requirements.
The image downloaded depends on type of machine:  [armv7l]( https://static.media.plabric.com/plugin-docker-image/armv7l/plabric-api-image-arm-version-v1-pv1.tar) [x86_64](https://static.media.plabric.com/plugin-docker-image/x86_64/plabric-api-image-x86-64-version-v1-pv1.tar
)

> * Installing Docker for the user _even with permission_ is not something I want to see a registered plugin doing, especially not if it then installs arbitrary unknown containers into said Docker instance. Hands off of system packages, there's SSH for this kind of stuff!

You are right, SSH would be the right way, but it's made that way to make things easier for people who doesn't know how to use SSH. I can remove it and keep the tutorial. ;)


> * Same goes for creating folders somewhere in the user's file system outside of the plugin folder ("docker_img").
> * Do NOT modify bundled KO bindings like `qrcode`, that will cause core OctoPrint and/or third party plugins to malfunction in really hard to debug ways!
> * Your plugin is not Python 3 compatible. Python 2 is EOL. OctoPrint 1.4.0 does support Python 2 and Python 3, plugins will be expected to do so as well (or at least support Python 3) during a transition period of roughly a year after the release of OctoPrint 1.4.0 to allow a migration of the whole ecosystem. Newly registered plugins should be Python 3 compatible out of the box. [See this forum thread for more information](https://community.octoprint.org/t/towards-python-3-and-octoprint-1-4-0/12382?u=foosel).

Ok, I ping you when changes are ready ;)

plabric commented:
Hi @foosel,

I have made  the following changes:

- I have removed the custom bindings for qrcode. Sorry, I didn't know it could affect to core Octoprint or another plugins. https://github.com/plabric/Octoprint-Plabric/commit/dccfaa7c4c887262c28b667b58f4909aee066243

- Now, the plugin uses the plugin's data folder to store the docker image. Image file will be deleted after installation. https://github.com/plabric/Octoprint-Plabric/commit/53f36f448199e76f54d424f28a32261287672ed6

- The plugin notifies the user, before install the requierements, that a file of 200Mb will be downloaded.
https://github.com/plabric/Octoprint-Plabric/commit/73e2b63c2324881261033f5754a248f8c7677c3d

- Now, the plugin is compatible with python3.
https://github.com/plabric/Octoprint-Plabric/commit/c1f6fc7be71b5b6d8b076dd3f960017e45922ab4
https://github.com/plabric/Octoprint-Plabric/commit/82362d08d626f1680e27cc0f209c5a7d7a8ab75e
https://github.com/plabric/Octoprint-Plabric/commit/626a64663951d79312cd64603258b2f6c2c79a01

- The code that installs automatically Docker is removed completely. Instead, the plugin offers a link to a tutorial depending on SO. 
https://github.com/plabric/Octoprint-Plabric/commit/9c393265cdc88dcd0830ffe94cf68f54c1d16ca0
https://github.com/plabric/Octoprint-Plabric/commit/476e321ce47fe5b2e4a9da5038c3080fd3a289b2
https://github.com/plabric/Octoprint-Plabric/commit/878b7f949de3cdeb04dd900ad483786a7d496510

- Plugin doesn't use the global API key anymore, instead, the docker image requests an app API Key using the Application Keys Plugin and following this workflow: https://docs.octoprint.org/en/master/bundledplugins/appkeys.html#sec-bundledplugins-appkey-datamodel-keyresponse
https://github.com/plabric/Octoprint-Plabric/commit/819e58a7b54acc9e12ee808b068c6634501a3cab
https://github.com/plabric/Octoprint-Plabric/commit/7817be5a3a9410e9393b6a18658b7b04db3ec6bd
https://github.com/plabric/Octoprint-Plabric/commit/7844d9b269f763ca80c24c8a88f9f61096627794


Tell me what you think about the changes and if I have to change something else. ;)

foosel commented:
Just a quick ping to let you know, I haven't forgotten you, just have my hands full with the 1.4.0 RCs. But I'll get back to this ASAP (hopefully this week!)

plabric commented:
Hi @foosel,

Congratulations on the new version of Octoprint with Python 3. It looks very good ;).
Seeing that now I can use libraries in Python3 I will simplify the Plugin code and will try to make the webrtc connection with python directly. So I will remove the dependencies with docker and not use a docker image. It will facilitate code review and reduce complexity for final users.

The Plugin will only work with Python3. There any problem?

I will ping you when changes are ready.

foosel commented:
1.4.0 took a bit of a toll, sorry for the ongoing delay.

plabric commented:
Hi @foosel,

Congratulations for Octoprint 1.4.0. ;) It looks very well.
Don't worry about time review, I'm working on remove all Docker dependencies and use pure python for simplify the plugin.
Now, with Python3 I've more options for create WebRTC connection without use a Docker image.

foosel commented:
Just keep in mind that most instances out there are still on python 2 and will probably continue to stay there for a while longer, as most of the plugin ecosystem isn't migrated yet.

foosel commented:
For the record, I've understood you that you are currently actively working on the plugin again, so that a review at present would not be sensible. Is that correct?

plabric commented:
Hi @foosel,

Sorry for the delay.
You're right, it's too early to launch pure Python3 plugin. 
Finally, the plugin works using Janus, all dependencies are packed in the plugin and it's not necessary the use of Docker anymore.

Thank you for your time.

foosel commented on _plugins/Plabric.md:
```diff
@@ -0,0 +1,65 @@
+
+## Support
+
+Do you have any doubt or trouble installing the plugin? Contact us at [contact@plabric.com](mailto:contact@plabric.com)
```
Minor correction, this should probably rather be "Do you have any question or problem" (emphasis mine). 

"Duda" en español puede ser "question" o "doubt", pero en este caso quieres que usar "question", "doubt" tiene una implicacion de desconfianza a tu producto.

plabric commented:
Hi @foosel,

Thank you for your time again ;)

I've moved the variables from the HTML to a new shared js file.
https://github.com/plabric/Octoprint-Plabric/commit/99e4c63bda05c6486b0cda6ae529a267399fe212
https://github.com/plabric/Octoprint-Plabric/commit/f8111262a44c9b8a303cce75c3b435ac009682e5
https://github.com/plabric/Octoprint-Plabric/commit/0fa786469d4d253ae501755992b30de19bac1993

I've modified the requests to disable, login and reconnect the plugin.
https://github.com/plabric/Octoprint-Plabric/commit/8eea2a0f391cc4398cd18e9363cfbe09223647bb
https://github.com/plabric/Octoprint-Plabric/commit/9f2a28ca99f02f9106dd2d2d2342a6e86b5dcc32

Empty less file has been removed.
https://github.com/plabric/Octoprint-Plabric/commit/4927159a36bc44a2ac85a884318acfcb3e14e474
https://github.com/plabric/Octoprint-Plabric/commit/5f3cfa216a6e7ddce76fe4227134b97aea017b74

Now, the video stream url is retrieved from settings.
https://github.com/plabric/Octoprint-Plabric/commit/0c917ef0fbf81e194151e80c9cb79f82633e57d3
https://github.com/plabric/Octoprint-Plabric/commit/e534d42ef30b9dcb7d1b7845cb1d1731008a7852
https://github.com/plabric/Octoprint-Plabric/commit/bffdac90d121d4cf1631252b467d02b091e4f41f
https://github.com/plabric/Octoprint-Plabric/commit/e40132309bdea215b94b480f6b174596c8703056
https://github.com/plabric/Octoprint-Plabric/commit/745f4c696de044905fc8cf8de4c902d23d44ea67
https://github.com/plabric/Octoprint-Plabric/commit/b2d438a7091f0b93be17c27dcd19cf6448f22eea

I've added info about video stream supported systems.
https://github.com/plabric/Octoprint-Plabric/commit/8decc688667dc201961040f2d8f00cf73e4520e3
https://github.com/plabric/plugins.octoprint.org/commit/6ce041387770a142ee61fc55c336218fc0620fe8

I've modified the contact info. Thank you for the advice ;)
https://github.com/plabric/Octoprint-Plabric/commit/95e29f51b5352f181546fa5c373ee875f1a50644
https://github.com/plabric/plugins.octoprint.org/commit/f663340b01c1a8757deaefa9f82997fa62116149
